src/synopsys-action/github-client-service.ts

```diff
@@ -0,0 +1,104 @@
+    this.repoOwner = process.env[GITHUB_ENVIRONMENT_VARIABLES.GITHUB_REPOSITORY_OWNER] || ''
+    this.githubServerUrl = process.env[GITHUB_ENVIRONMENT_VARIABLES.GITHUB_SERVER_URL] || ''
+    this.githubApiURL = this.githubServerUrl === constants.GITHUB_CLOUD_URL ? constants.GITHUB_CLOUD_API_URL : this.githubServerUrl
+    this.commit_sha = process.env[GITHUB_ENVIRONMENT_VARIABLES.GITHUB_SHA] || ''
+    this.githubRef = process.env[GITHUB_ENVIRONMENT_VARIABLES.GITHUB_REF] || ''
+  }
+
+  async uploadSarifReport(defaultSarifReportDirectory: string, userSarifFilePath: string): Promise<void> {
+    info('Uploading SARIF results to GitHub')
+    if (isNullOrEmptyValue(inputs.GITHUB_TOKEN)) {
+      throw new Error('Missing required GitHub token for uploading SARIF report to GitHub Advanced Security')
+    }
+    let retryCountLocal = RETRY_COUNT
+    let retryDelay = RETRY_DELAY_IN_MILLISECONDS
+    const stringFormat = (url: string, ...args: string[]): string => {
+      return url.replace(/{(\d+)}/g, (match, index) => args[index] || '')
+    }
+    const endpoint = stringFormat(this.githubApiURL.concat(this.gitHubCodeScanningUrl), this.repoOwner, this.repoName)
+    const sarifFilePath = userSarifFilePath ? userSarifFilePath : getDefaultSarifReportPath(defaultSarifReportDirectory, true)
+
+    if (checkIfPathExists(sarifFilePath)) {
+      try {
+        const sarifContent = fs.readFileSync(sarifFilePath, 'utf8')
+        const compressedSarif = zlib.gzipSync(sarifContent)
+        const base64Sarif = compressedSarif.toString('base64')
+        const data = {
+          commit_sha: this.commit_sha,
+          ref: this.githubRef,
+          sarif: base64Sarif,
+          validate: true
+        }
+        do {
+          const httpClient = new HttpClient('GithubClientService')
+          const httpResponse = await httpClient.post(endpoint, JSON.stringify(data), {
+            Authorization: `Bearer ${this.githubToken}`,
+            Accept: 'application/vnd.github+json'
+          })
+          debug(`HTTP Status Code: ${httpResponse.message.statusCode}`)
+          debug(`HTTP Response Headers: ${JSON.stringify(httpResponse.message.headers)}`)
+          const responseBody = await httpResponse.readBody()
+          const rateLimitRemaining = httpResponse.message?.headers['x-ratelimit-remaining'] || ''
+          if (httpResponse.message.statusCode === 202) {
+            info('SARIF result uploaded successfully to GitHub Advance Security')
+            retryCountLocal = 0
+          } else if (httpResponse.message.statusCode === 403 && (rateLimitRemaining === '0' || responseBody.includes('secondary rate limit'))) {
+            retryDelay = await this.retrySleepHelper('Uploading SARIF report to GitHub Advanced Security has been failed due to rate limit, Retries left: ', retryCountLocal, retryDelay)
+            if (retryCountLocal === 1) {
+              const rateLimitReset = httpResponse.message?.headers['x-ratelimit-reset'] || ''
```

Comment:
Shall we check for reteLimitReset is more than retrial time period. If rateLimitReset time is more, we can skip retry

Reply:
if reset time is <= 115 sec then only we will retry, otherwise show error message with x minutes left